### PR TITLE
[scheduler] bucketed scheduler - single ticker implementation

### PR DESCRIPTION
**Pull request by truthbk**
### What does this PR do?

Introduces the bucketed scheduler with a simpler single ticker implementation. It will not enforce deadlines and will stay at a bucket until all checks in it have been correctly enqueued. We lose some flexibility in unlikely scenarios where we might be running low on runners, but gain predictability.

### Motivation

We want to spread the load during collection intervals, hence the buckets. The single ticker is aimed at providing a simple, predictable approach to the scheduling of the checks in these buckets. 

### Additional Notes

Anything else we should know when reviewing?

### CODE REVIEW

**Comment by truthbk** [comment]
Not sure why the reno linter is failing 🤷‍♂️

**Comment by arbll** [comment]
Some logic was added to enforce team labels. We should probably rename it / split the test in two

**Comment by arbll** [comment]
I feel that we need more visibility on what is going on in the scheduler/workers to help with identifying and debugging potential issues.
A few ideas : 
1. for this specific PR, even if we don't enforce the deadline, logging when we are behind the schedule would be valuable
2. dumping the state of the workers when we reach the deadline (what check are run and for how long)

Not sure if 2. completely fits the scope of this task, your call.

**Comment by truthbk** [comment]
First off, I'd like to stress there is no such concept as a deadline with the single ticker implementation. The 15 second interval is a best effort thing, the interval can last as long as it takes, going over the 15 buckets - this is the main reason I originally implemented the multi-ticker approach. 

The best we can really do is check if we've fallen behind the expected collection interval (ie. were unable to schedule all checks before the next tick). I will log something there. 

Regarding what checks are being run and for how long is useful but I believe is in the scope of a separate PR. That said, I believe the logic is quite clear, all checks will be run by definition, if they're not, we've lost all runners and our scheduler has stalled.

File: pkg/collector/scheduler/job.go
Comment (arbll) [inline pkg/collector/scheduler/job.go]: First time I see this algorithm but it should be a good addition.
I think we have an unhandled edge case with `2` : https://play.golang.org/p/8aPXsdpVxzY
Do you have examples of other implementations of thins in case we missed anything else ?

File: pkg/collector/scheduler/job.go
Comment (arbll) [inline pkg/collector/scheduler/job.go]: Does it make sense to add by how much we missed ?

File: pkg/collector/scheduler/job.go
Comment (arbll) [inline pkg/collector/scheduler/job.go]: I might be missing something obvious bu shouldn't this be `jq.lastTick.Add(1*time.Second)` ?

**Review by arbll — COMMENTED** [review]
Thanks for adding some logging. This implementation is indeed straightforward and 2. might be overkill here.

This should be the last batch of feedback before I approve and we can test this in real life scenarios 🚀

File: pkg/collector/scheduler/job.go
Comment (truthbk) [reply pkg/collector/scheduler/job.go]: The first problem is ticks aren't 100% precise. So when I tried this with 1 second, there were cases when we had slept just slightly more than a second. Prompting an incorrect log message, we were not running behind.

Thus, I considered it a missed scheduling slot if |t1| never fired, which would mean approximately two seconds between t0 and t2.
 
```
....----|t0|-----SCHED----|t1|-----SCHED-----|t2|----.... 
```

Another option might be to try to play the game of estimating just how precise those tickers can get 1ms, 10us, etc... and use that as the threshold. I didn't really want to play that fine-tuning game hence the 2sec (because that implied you had clearly missed your bucketing schedule slot).

File: pkg/collector/scheduler/job.go
Comment (truthbk) [reply pkg/collector/scheduler/job.go]: I didn't base myself off any other implementation. I deduced this from modular arithmetic. The goals were:
1) let's schedule the checks as far apart from each other as possible, so as to smoothen the CPU load over the entire interval (best effort).
2) make sure that the steps would make us fill every bucket with a single check, before adding the next round of checks. 

If you need formal mathematical proof I can work on it. But basically the above will hold true (we could improve and make it even more sparse, but I don't think we need to go crazy) when:
a) Even number of buckets: we use the first odd index to the left of the middle index (doesn't matter if that index is even or odd - ie. 20 -> 9, 22 -> 9). 
b) Odd number of buckets: we use the middle index. (ie. 9 -> 4, 15 -> 7).

The reason this works is because when we wrap around the modular ring, we "land" in a bucket that was previously unused. With case (b) this is quite easy to see, we are always one off the previous location when wrapping around. (ie. 9: 0, 4, 8, 3, 7, 2, 6, 1, 5). With case (a) it's less trivial, you need to rely on the fact that since the number of buckets is even, and our "step" is odd, in the general case we will have three "steps" before wrapping around and since `odd*odd = odd`, so by definition we will not land on the same bucket, and the offset is such that we will continue to wrap around landing in free buckets until all buckets have been filled.

Note:  for (a) if N <= 6, the general case begins to break down, as we go down to a step of just 1, but that will work just fine as well. It's easy to see that steps of 2 or 3 would not work as we'd always be hitting the same buckets.

File: pkg/collector/scheduler/job.go
Comment (truthbk) [reply pkg/collector/scheduler/job.go]: I don't think so, `interval = 0` is a long-running check, and is handled differently here: https://github.com/DataDog/datadog-agent/blob/master/pkg/collector/scheduler/scheduler.go#L67-L70

File: pkg/collector/scheduler/job.go
Comment (truthbk) [reply pkg/collector/scheduler/job.go]: Added it.

**Review by arbll — DISMISSED** [review]
Let's merge this bad boy

File: releasenotes/notes/bucketed_scheduler-d4eca4b9e20edc67.yaml
Comment (truthbk) [reply releasenotes/notes/bucketed_scheduler-d4eca4b9e20edc67.yaml]: Yeah you might be right, I might change this. Thanks buddy! 👌

**Comment by truthbk** [comment]
Previously approved, review went stale after updating release note. Merging.

**Pull request merged**
